### ISSUE TITLE
Crash in vexocide/5406

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#5581]
I got this while sending what appear to me to be well-formed requests:

```
error: Error in /usr/local/include/boost/optional/optional.hpp at line 630:
error: BOOST_ASSERT failure in 'reference_const_type boost::optional<auth::user_t>::get() const': this->is_initialized()
error: Backtrace:
error: Wed Mar 30 17:34:41 2016

       1 [0x20d88cb]: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:108
       2 [0x1b21920]: backtrace_t::backtrace_t() at backtrace.cc:210 (discriminator 1)
       3 [0x1b22cdc]: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:331
       4 [0x1b215a7]: format_backtrace(bool) at backtrace.cc:205
       5 [0x209dfde]: report_fatal_error(char const*, int, char const*, ...) at errors.cc:85
       6 [0x209e74a]: boost::assertion_failed(char const*, char const*, char const*, long) at errors.cc:312
       7 [0x186889e]: boost::optional<auth::user_t>::get() const at optional.hpp:630 (discriminator 8)
       8 [0x186a977]: /home/ssd2/mlucy/rethinkdb/build/debug_clang/rethinkdb() [0x186a977] at 0x186a977 ()
       9 [0x186a692]: /home/ssd2/mlucy/rethinkdb/build/debug_clang/rethinkdb() [0x186a692] at 0x186a692 ()
       10 [0x1e5f038]: std::function<void (auth_semilattice_metadata_t const*)>::operator()(auth_semilattice_metadata_t const*) const at functional:2471
       11 [0x1e5efa3]: void cross_thread_watchable_variable_t<auth_semilattice_metadata_t>::apply_read<std::function<void (auth_semilattice_metadata_t const*)> const&>(std::function<void (auth_semilattice_metadata_t const*)> const&) at cross_thread_watchable.hpp:38
       12 [0x1e5eef1]: cross_thread_watchable_variable_t<auth_semilattice_metadata_t>::w_t::apply_read(std::function<void (auth_semilattice_metadata_t const*)> const&) at cross_thread_watchable.hpp:56
       13 [0x18693e4]: auth::scram_authenticator_t::first_message(std::string const&) at scram_authenticator.cc:59 (discriminator 2)
       14 [0x20a8b4e]: query_server_t::handle_conn(scoped_ptr_t<linux_tcp_conn_descriptor_t> const&, auto_drainer_t::lock_t) at server.cc:410 (discriminator 2)
```

**Comment by mlucy** [comment]
Actually, this appears to happen if I close the connection in the middle of a valid request, just after receiving the message from step 2.  It may also happen in other circumstances, I haven't investigated too carefully.

**Comment by mlucy** [comment]
Here are the two messages I sent that caused this:
```
"\xC3\xBD\xC24"
"{\"protocol_version\":0,\"authentication_method\":\"SCRAM-SHA-256\",\"authentication\":\"n,,n=admin,r=g1g0+m9EBCvuCZ8Rxvg7edWL\"}\0"
```

I'm gonna go look at the Python implementation again, it's possible I misread it.

**Comment by mlucy** [comment]
Ah, looking at the code it seems like the case where a user doesn't have a password (like the "admin" user) just isn't implemented yet.

**Comment by VeXocide** [comment]
@mlucy -- could you please check which branch you're on, because this shouldn't be the case in `vexocide/5406` but was in `vexocide/4519`?

**Comment by mlucy** [comment]
@VeXocide -- I'm on `vexocide/5406`.  The exact commit is `f02637f`.  It's possible something weird happened, I could rebuild from scratch and try to reproduce if you want.

**Comment by VeXocide** [comment]
This really shouldn't have happened at this point, it'd be great if you could find a way to reproduce this.

**Comment by mlucy** [comment]
I decided to just try reproducing on `next` to save us some time, and this crash appears not to happen there, so it was either a build problem or was fixed between `f02637f` and `next`.
